srk/src/numberField.mli

```diff
@@ -0,0 +1,214 @@
+open Polynomial
+
+(** Given a univariate polynomial, and a variable dimension, create an
+    equivalent multivariate polynomial using the given dimension.*)
+val make_multivariate : Monomial.dim -> QQX.t -> QQXs.t
+
+(** Given a multivariate polynomial over a single dimension, create an
+    equivalent univariate polynomial.
+    @raise Failure if input is not a univariate polynomial.*)
+val make_univariate : QQXs.t -> QQX.t
+
+(** Let K1 = Q[v0,v1]/p and K2=Q[v0,v1]/q be number fields. Let K1 have degree d1 and K2 have degree d2.
+    [prim, v0_in_prim, v1_in_prim] = [primitive_elem (d1*d2) p q v0 v1] 
+    is such that the number field K3 = Q[x]/prim of degree d1*d2 such that K3 contains K1 and K2. 
+    [v0_in_prim] and [v1_in_prim] give the representations of v0 and v1 in K3 respectively. For
+    the method to work p needs to be irreducible in K2 and q needs to be irreducible in K1.*)
+val primitive_elem : int -> QQXs.t -> QQXs.t -> Monomial.dim -> Monomial.dim -> QQX.t * QQX.t * QQX.t
+
+
+module type NF = sig  
+
+  (** The degree of the extension.*)
+  val deg : int
+
+  (** int_poly is a monic integer polynomial that defines the same field as A.min_poly. *)
+  val int_poly : QQX.t
+
+  (** The type of elements of the number field. *)
+  type elem
+
+  (** Converts a univariate polynomial to an element of the number field.*)
+  val make_elem : QQX.t -> elem
```

Comment:
See earlier comment for a distinction between A.min_poly and int_poly. Also, I'm not sure I exactly understand what you're saying. First, remember that the minimal poly does *not* factor in the number field. You will get at least two factors, but not more than that. So at this point we do not know the factors of either int_poly or A.min_poly. To figure them out we need to compute a splitting field first, then factor. R.e to your first comment. Let alpha be the root of  A.min_poly corresponding to x. For the implementation of the order which comes later, it is convenient to have a algebraic integer in the number field. At instantiation we compute such an integer as follows. If A.min_poly is a monic integer polynomial, then we take the algebraic integer alpha, and int_poly = A.min_poly. If A.min_poly is not a monic integer polynomial we compute the lcm of the dens of A.min_poly. Then beta = lcm*alpha is an algebraic integer. int_poly is then the minimal polynomial of beta. Then internally elements of the number field are represented as univariate polynomials modulo the minimal polynomial of beta, say q(x). The purpose of make_elem is that externally, the user supplied A.min_poly <> q(x). And thus assumes the field is Q[x]/A.min_poly. But this is not the case, it's Q[x]/q(x). Make_elem recognizes the isomorphism and performs the conversion. E.x. if A.min_poly = x^2+1, then make_elem x = x. I.e. x is represented as x. However, if A.min_poly = x^2 + 1/4, then int_poly = x^2 + 4. Then make_elem x = 1/4 x. Externally, x is either -i/2 or i/2 and for any computations of NF it doesn't matter which you choose. Internally, x corresponds to either -2i or 2i. Again it doesn't matter which one you use.

Reply:
My basic confusion is that I wouldn't expect the client to know that internally Q(a) is represented by QQ[x]/p (where p is the supplied minimal polynomial of a or some other polynomial). But I think it's ok if we do want to expose that representation.

Reply:
For computing algebraic relations of exponential polynomials in rational.ml I need to get ahold of some representation of field elements as polynomials. This is because we get algebraic relations over only rational numbers by creating an ideal I involving polynomials p1(x), ..., pk(x) \in Q[x]/q for some q. p1, ..., pk represent number field elements we get purely rational relations by creating <I, p1(x), ..., pk(x), q(x)> and eliminating x. There is a question of whether the user should be able to supply number field elements as polynomials. This is the purpose of make_elem. I use make_elem only in test_numberfield.ml so functionally it's not really important. There currently is a mismatch between make_elem and get_poly though. Because int_poly does not necessarily equal A.min_poly, get_poly (make_elem p) does not necessarily equal p. I at least need some version of get_poly, and it is more convenient to just get the internal representation of a number field element. On the other hand make_elem p computes a number field element assuming p is and element of Q[x]/A.min_poly because A.min_poly is the defining polynomial the user supplied. I could (1) remove make_elem, (2) make make_elem be with respect to int_poly rather than A.min_poly, (3) make get_poly be with respect to A.min_poly rather than int_poly, or (4) leave it as is.